vernac/declare.ml

```diff
@@ -2273,8 +2273,15 @@ let solve_and_declare_by_tac prg obls i tac =
   | Some (t, ty, uctx) ->
     let obl = obls.(i) in
     let poly = Internal.get_poly prg in
-    let prg = ProgramDecl.Internal.set_uctx ~uctx prg in
+    let uctx = if poly then uctx else UState.union prg.prg_uctx uctx in
     let def, obl', _cst = declare_obligation prg obl ~body:t ~types:ty ~uctx in
+    let uctx =
+      (* imitate obligation_terminator, XXX no idea if this is really correct *)
+      if poly then UState.union prg.prg_uctx uctx
+      else if def then UState.Internal.reboot (Global.env()) uctx
+      else uctx
+    in
```

Comment:
The `set_uctx` function is pure and we're calling it a couple of lines below with a similar dispatch on poly / def, can't we merge both calls?

Reply:
@SkySkimmer ping

Reply:
I don't know

Reply:
What do you mean, I don't know? You can clearly simplify the code.

Reply:
right, I forgot how this code looked